File: sdk/ruby/lib/rocksky/resources/player.rb
module Rocksky
  module Resources
    # `app.rocksky.player.*` endpoints — remote control of a Rocksky player.
    class Player < Base
      # Currently playing track.
      def get_currently_playing(player_id: nil, actor: nil)
        query("app.rocksky.player.getCurrentlyPlaying",
              playerId: player_id, actor: actor)
      end

      # Playback queue.
      def get_playback_queue(player_id: nil)
        query("app.rocksky.player.getPlaybackQueue", playerId: player_id)
      end

      # Resume playback.
      def play(player_id: nil)
        procedure("app.rocksky.player.play", params: { playerId: player_id })
      end

      # Pause playback.
      def pause(player_id: nil)
        procedure("app.rocksky.player.pause", params: { playerId: player_id })
      end

      # Skip to next track.
      def next(player_id: nil)
        procedure("app.rocksky.player.next", params: { playerId: player_id })
      end

      # Go to previous track.
      def previous(player_id: nil)
        procedure("app.rocksky.player.previous", params: { playerId: player_id })
      end

      # Seek to position in milliseconds.
      def seek(position:, player_id: nil)
        procedure("app.rocksky.player.seek",
                  params: { playerId: player_id, position: position })
      end

      # Play a single file.
      def play_file(file_id:, player_id: nil)
        procedure("app.rocksky.player.playFile",
                  params: { playerId: player_id, fileId: file_id })
      end

      # Play a directory.
      def play_directory(directory_id:, player_id: nil, shuffle: nil,
                         recurse: nil, position: nil)
        procedure("app.rocksky.player.playDirectory",
                  params: {
                    playerId: player_id,
                    directoryId: directory_id,
                    shuffle: shuffle,
                    recurse: recurse,
                    position: position
                  })
      end

      # Append items to the queue. `items` is an Array of file identifiers.
      def add_items_to_queue(items:, player_id: nil, position: nil, shuffle: nil)
        procedure("app.rocksky.player.addItemsToQueue",
                  params: {
                    playerId: player_id,
                    items: items,
                    position: position,
                    shuffle: shuffle
                  })
      end
    end
  end
end

File: sdk/python/src/rocksky/resources/player.py
"""``app.rocksky.player.*`` — remote playback control (Rockbox/RockboxJS)."""

from __future__ import annotations

from typing import Any

from ._base import Resource


class PlayerResource(Resource):
    async def currently_playing(
        self,
        *,
        player_id: str | None = None,
        actor: str | None = None,
    ) -> Any:
        return await self._transport.query(
            "app.rocksky.player.getCurrentlyPlaying",
            params={"playerId": player_id, "actor": actor},
        )

    async def queue(self, *, player_id: str | None = None) -> Any:
        return await self._transport.query(
            "app.rocksky.player.getPlaybackQueue",
            params={"playerId": player_id},
        )

    async def play(self, *, player_id: str | None = None) -> Any:
        return await self._transport.procedure(
            "app.rocksky.player.play", params={"playerId": player_id}, auth=True
        )

    async def pause(self, *, player_id: str | None = None) -> Any:
        return await self._transport.procedure(
            "app.rocksky.player.pause", params={"playerId": player_id}, auth=True
        )

    async def next(self, *, player_id: str | None = None) -> Any:
        return await self._transport.procedure(
            "app.rocksky.player.next", params={"playerId": player_id}, auth=True
        )

    async def previous(self, *, player_id: str | None = None) -> Any:
        return await self._transport.procedure(
            "app.rocksky.player.previous", params={"playerId": player_id}, auth=True
        )

    async def seek(self, position: int, *, player_id: str | None = None) -> Any:
        return await self._transport.procedure(
            "app.rocksky.player.seek",
            params={"playerId": player_id, "position": position},
            auth=True,
        )

    async def play_file(self, file_id: str, *, player_id: str | None = None) -> Any:
        return await self._transport.procedure(
            "app.rocksky.player.playFile",
            params={"playerId": player_id, "fileId": file_id},
            auth=True,
        )

    async def play_directory(
        self,
        directory_id: str,
        *,
        player_id: str | None = None,
        shuffle: bool | None = None,
        recurse: bool | None = None,
        position: int | None = None,
    ) -> Any:
        return await self._transport.procedure(
            "app.rocksky.player.playDirectory",
            params={
                "playerId": player_id,
                "directoryId": directory_id,
                "shuffle": shuffle,
                "recurse": recurse,
                "position": position,
            },
            auth=True,
        )

    async def add_items_to_queue(
        self,
        items: list[str],
        *,
        player_id: str | None = None,
        position: int | None = None,
        shuffle: bool | None = None,
    ) -> Any:
        return await self._transport.procedure(
            "app.rocksky.player.addItemsToQueue",
            params={
                "playerId": player_id,
                "items": items,
                "position": position,
                "shuffle": shuffle,
            },
            auth=True,
        )

    async def add_directory_to_queue(
        self,
        directory: str,
        *,
        player_id: str | None = None,
        position: int | None = None,
        shuffle: bool | None = None,
    ) -> Any:
        return await self._transport.procedure(
            "app.rocksky.player.addDirectoryToQueue",
            params={
                "playerId": player_id,
                "directory": directory,
                "position": position,
                "shuffle": shuffle,
            },
            auth=True,
        )
